File: aula-07/vetor/java/src/main/java/com/github/caiobas/es/cs/aula07/domain/NumeroFrequente.java
package com.github.caiobas.es.cs.aula07.domain;

import java.util.Random;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

/**
 * Classe com o objetivo de retornar o número com a
 * maior quantidade de repetições de 1.000.000 de números
 * sorteados aleatoriamente entre o intervalo de 0 a 1.000.
 */
public final class NumeroFrequente {

    private NumeroFrequente() {

    }
    /**
     * Classe com o objetivo de gerar 1.000.000 de números
     * no intervalo de 0 a 1.000 aleatoriamente.
     *
     * @param quantidade Quantidade de números gerados.
     * @param intervalo Intervalo dos números gerados.
     *
     * @return Retorna um ArrayList de inteiro com
     * todos os números gerados aleatoriamente.
     *
     * @throws IllegalArgumentException Se a quantidade não for 1.000.000.
     * @throws IllegalArgumentException Se o intervalo não for 1.000.
     */
    public static ArrayList<Integer> numerosAleatorios(final int quantidade, final int intervalo) {
        final int limiteQuantidade = 1000000;
        if (quantidade != limiteQuantidade) {
            throw new IllegalArgumentException("Quantidade deve ser 1000000.");
        }
        final int limiteIntervalo = 1000;
        if (intervalo != limiteIntervalo) {
            throw new IllegalArgumentException("Intervalo deve ser entre 0 e 1000.");
        }

        Random random = new Random();

        ArrayList<Integer> vetor = new ArrayList<Integer>();

        final int intervaloFinal = intervalo + 1;

        for (int contador = 1; contador <= quantidade; contador++) {
            vetor.add(random.nextInt(intervaloFinal));
        }

        return vetor;
    }

    /**
     * Classe com o objetivo analisar o ArrayList de
     * números gerados aleatoriamente e retornar um
     * ArrayList com o(s) número(s) mais frequentes.
     *
     * @param vetor Vetor com números gerados aleatoriamente.
     *
     * @return Retorna um ArrayList de inteiro com os
     * números mais frequentes do ArrayList de aleatórios.
     */
    public static ArrayList<Integer> obterNumeroFrequente(final ArrayList<Integer> vetor) {
        HashMap<Integer, Integer> numeros = new HashMap<Integer, Integer>();

        for (int inteiro : vetor) {
            if (numeros.containsKey(inteiro)) {
                numeros.put(inteiro, numeros.get(inteiro) + 1);
            } else {
                numeros.put(inteiro, 1);
            }
        }

        int compara = 0;
        for (int inteiro : numeros.keySet()) {
            if (numeros.get(inteiro) > compara) {
                compara = numeros.get(inteiro);
            }
        }

        ArrayList<Integer> frequentes = new ArrayList<Integer>();
        for (int inteiro : numeros.keySet()) {
            if (numeros.get(inteiro) == compara) {
                frequentes.add(inteiro);
            }
        }

        Collections.sort(frequentes);
        return frequentes;

    }

}

File: aula-07/vetor/js/NumeroFrequente.js
/**
 * Função com o objetivo de analisar um Array de
 * 1.000.000 números no intervalo de 0 a 1.000 gerados
 * aleatoriamente e retornar um Array com o(s) número(s) mais frequentes.
 * 
 * @param {number} quantidade Quantidade de números gerados.
 * @param {number} intervalo Intervalo dos números gerados.
 * 
 * @returns {Array} Retorna um Array de inteiros com o(s)
 * número(s) mais frequentes do Array de aleatórios.
 * 
 * @throws {TypeError} Se quantidade não for um número.
 * @throws {TypeError} Se intervalo não for um número.
 * @throws {RangeError} Se quantidade não for 1.000.000.
 * @throws {RangeError} Se intervalo não for 1.000.
 */
function numeroFrequente(quantidade, intervalo) {

    if(typeof quantidade != "number") {
        throw new TypeError("Quantidade deve ser número.");
    }

    if(typeof intervalo != "number") {
        throw new TypeError("Intervalo deve ser número.");
    }
    
    if(quantidade != 1000000) {
        throw new RangeError("Quantidade deve ter valor de 1.000.000.");
    }

    if(intervalo != 1000) {
        throw new RangeError("Intervalo deve ter valor de 1000.");
    }

    let vetor = new Array();

    for(let contador = 0; contador < quantidade; contador++){
        vetor[contador] = Math.floor(Math.random() * (intervalo + 1));
    }
    vetor.sort();

    let numeros = new Map();

    for(let contador of vetor){
        if(numeros.get(contador)){
            numeros.set(contador, numeros.get(contador) + 1);
        } else {
            numeros.set(contador, 1);
        }
    }

    let compara = 0;

    for(let contador of vetor){
        if(numeros.get(contador) > compara){
            compara = numeros.get(contador);
        }
    }

    let frequentes = new Array();
    let contador1 = 0;
    
    for(let contador of vetor){
        if(numeros.get(contador) == compara && frequentes[contador1 - 1] != contador){
            frequentes[contador1] = contador;
            contador1++;
        }
    }

    frequentes.sort();
    console.log(frequentes);
    return frequentes;
}

module.exports = numeroFrequente;